### ISSUE TITLE
How to define correlation coefficient as a custom metric function

### ISSUE THREAD

**Issue by TrifonT** [Microsoft/CNTK#3314]
For a simple network with one output only, I want to correlate output with the target. How should I define a custom metric function for this?

**Comment by delzac** [comment]
Correlation coefficient is calculated across your entire dataset right? I think  implementing [this](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient
) formula is what you are looking for.

I think you will be better served by posting your question on stackoverflow as its not related to cntk. You will get a faster response.

**Comment by TrifonT** [comment]
@delzac thank you for your answer !
I need to define it as a Function and pass it as a parameter to CNTK.Trainer class. 

I have tried the following definition:
```
def corr_coef(output, target):
    dtype = get_data_type(output, target)
    output = sanitize_input(output, dtype)
    target = sanitize_input(target, dtype)
    return np.corrcoef(output, target)[0,1]
```
but this won't work, I get the error 'Variable' object has no attribute 'conjugate'

**Comment by delzac** [comment]
Hi,

Are you going to use it as a metric or a loss function? If it is a metric, its simpler if you calculate it outside of CNTK.

If you intend to use it as a loss function (i.e. you intend to take its gradient), then you will have to implement it used cntk and numpy cannot be used to build the graph.

Your function "corr_coef" doesn't work as numpy cannot be used to build graphs.

**Comment by delzac** [comment]
Anyway, are you looking for this function C.unpack_batch? It shift the dynamic batch axis into a static axis and you can calculate your mean from there. If you intend to calculate corr_coeff natively.

**Comment by TrifonT** [comment]
Thanks again. I need it as a metric function, just to measure the fitness of output versus target. 

I have tried the following:

```
def corr_coef(output, target):
    xm = C.mean(output)
    ym = C.mean(target)
    x = C.minus(output, xm);
    y = C.minus(target, ym);
    xy = C.ops.times_transpose(x,y)
    sx2 = C.ops.times_transpose(x,x)
    sy2 = C.ops.times_transpose(y,y)
    
    return  xy /   ( C.ops.sqrt( sx2 ) * C.ops.sqrt( sy2 ) )
```
but it doesn't seem to work. I will look about the C.unpack_batch function.

**Comment by delzac** [comment]
If you need it only as a metric then its much much easier to just take the

output_value = model.eval()  # This is a list of numpy array
and feed output_value into numpy.

Yup, your function definition needs an unpack_batch first, then calculate the mean.
